recipes-core/initramfs-framework/files/composefs

```diff
@@ -10,14 +14,105 @@ composefs_error() {
 	fatal "$@"
 }
 
+# We could replace this whole function by:
+#
+# $ ostree admin post-copy --sysroot="${SYSROOT_DIR}"
+#
+# However this would bring in the dependency of ostree to the ramdisk
+# which doesn't look like a good idea. Instead, here we depend only on
+# "findutils" and "fsverity-utils".
+#
+composefs_enable_fsverity_all() {
+	# Enable verity in all repository objects.
+	find "${SYSROOT_DIR}/ostree/repo/objects" -exec fsverity enable '{}' \; 2>/dev/null
+
+	# And also on the composefs file of every deployment.
+	for cfsfile_ in "${SYSROOT_DIR}/ostree/deploy"/*/deploy/*/"${COMPOSEFS_NAME}"; do
+		fsverity enable "${cfsfile_}" 2>/dev/null
+	done
+}
```

Comment:
I saw from the commit message that on first boot fsverity can take quite a bit of time on first boot. If the alternative method with `ostree`, as mentioned in the comment, was used instead would this reduce the time? Or not really?

Furthermore, is there any possible optimizations that could be done with regards to this setup time on first boot? Other than adding knowledge of fsverity to the installer and build machine like your commit message stated.

Reply:
The operations done here are exactly the same done by `ostree post-copy` and all the time is taken by the kernel to determine the Merkle tree for each file. Considering that, the only possibility to optimize the total time I can imagine would be trying to parallelize the operation over the various CPU cores. But even this may not help because I see a message saying that the kernel is using the CAAM for determining the hashes so unless the CAAM itself also has independent hashing units, the parallelization may not help.

Thus, if this extra time becomes a problem, I think the best solution would be moving the fsverity enabling to Toradex Easy Installer because then we at least wouldn't need the information to be part of the image (which would require changing the image format, today a simple tarball).

The other option of creating an image with fsverity information would be a big change since it would impact the installer and complicate tooling such as TCB. And it would require the build machine to support fsverity as well, i.e. it'd put constraints on user's machines (I mean those building with Yocto and those using TCB). So, I guess this would be the worst solution.

Reply:
I'm not sure if I'm solely capable of deciding whether this initial boot time is a problem or not. In the end it's up to the users. I know some users are quite sensitive regarding boot-time and all, even for one time stuff like this. I suppose they could just do the initial boot in production after flashing so that this initial boot time is not experienced when the device is actually deployed. Though this could also be a complication in terms of production workflows for some users, needing to boot their systems and wait for the initialization to complete on the production floor.

Yeah it's a complex problem with no easy solution I see. In the end it's not a "blocking issue" or anything but certainly something to be aware of going forward.

Reply:
Absolutely. Do you think we should ask for PM feedback at this point?

Reply:
Getting more opinions on the matter wouldn't hurt.